consensus/polybft/fsm.go

```diff
@@ -429,10 +429,29 @@ func (f *fsm) VerifyStateTransactions(transactions []*types.Transaction) error {
 func (f *fsm) Insert(proposal []byte, committedSeals []*messages.CommittedSeal) (*types.FullBlock, error) {
 	newBlock := f.target
 
+	var proposedBlock types.Block
+	if err := proposedBlock.UnmarshalRLP(proposal); err != nil {
+		return nil, fmt.Errorf("failed to insert proposal, block unmarshaling failed: %w", err)
+	}
+
+	if newBlock == nil || newBlock.Block.Hash() != proposedBlock.Hash() {
```

Comment:
Should we return error immediately in case `cachedBlock` and `proposedBlock` mismatch?

```suggestion
        if newBlock.Block.Hash() != proposedBlock.Hash() {
             return errors.New("failed to insert proposal: cached block and proposed blocks mismatch")
        }
	if newBlock == nil  {
```

Reply:
In my opinion - Yes. If the proposal you validated is not equal to the proposal you got to insert, then some malicious behavior is possibly at work, or a bug on client side. Better let syncer insert the block, and make sure we didn't mess something up, just to be sure.

Reply:
I would do same as Goran suggested

Reply:
if `f.target` is `nil` i would return from function immediately with error

Reply:
We agreed to return an error here.